hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/store/AbstractNearCacheRecordStore.java

```diff
@@ -412,13 +404,25 @@ public long tryReserveForUpdate(K key, Data keyData) {
             return NOT_RESERVED;
         }
 
-        R reservedRecord = getOrCreateToReserve(key, keyData);
         long reservationId = nextReservationId();
-        if (reservedRecord.casRecordState(RESERVED, reservationId)) {
-            return reservationId;
-        } else {
+        R reservedRecord = getOrCreateToReserve(key, keyData, reservationId);
+        return reservedRecord.getReservationId() == reservationId ? reservationId : NOT_RESERVED;
+    }
+
+    @Override
+    public long tryReserveForCacheOnUpdate(K key, Data keyData) {
+        checkAvailable();
+        // if there is no eviction configured we return if the Near
+        // Cache is full and it's a new key (we have to check the
+        // key, otherwise we might lose updates on existing keys)
+        if (evictionDisabled && evictionChecker.isEvictionRequired() && !containsRecordKey(key)) {
             return NOT_RESERVED;
         }
+
+        long reservationId = nextReservationId();
+        R reservedRecord = reserveForCacheOnUpdate(key, keyData, reservationId);
+        return reservedRecord == null ? NOT_RESERVED
```

Comment:
CACHE_ON_UPDATE mode has different characteristics than INVALIDATE mode when updating local near-cache. During update, if CACHE_ON_UPDATE finds a previously reserved record, it is deleted. This is different from INVALIDATE mode which doesn't delete previously reserved record and keeps it as is. The reason for this deletion is: concurrent reservation attempts. If CACHE_ON_UPDATE doesn't delete previously reserved record, indefinite read of stale value situation can be seen. Since we don't apply invalidations which are sent from server to near-cache if the source UUID of the invalidation is same with the end's UUID which has near-cache on it (client or server UUID which has near cache on it).

Reply:
Thanks!